File: homework2/Aditya_Prasad/q5_stack.cpp
// Time taken : 10 mins

#include <bits/stdc++.h>
using namespace std;

struct Node{
    int data;
    Node* next;

    Node(int val){
        data = val;
        next = nullptr;
    }
};

class Stack {
private:
    Node* front;

public:
    Stack(){
        front = nullptr;
    }
    int top(){
        return front->data;
    } // returns the top item in the stack. O(1) time.
    void push(int x){
        Node* newNode = new Node(x);
        newNode->next = front;
        front = newNode;
    } // adds x to the top of the stack. O(1) time.
    int pop(){
        Node* temp = front;
        int topVal = front->data;
        front = front->next;
        delete temp;
        return topVal;
    }// removes and returns the top item in the stack. O(1) time.
    bool isEmpty(){
        if(front == nullptr) return true;
        else return false;
    }// returns a boolean indicating whether the stack is empty. O(1) time.
}; 


int main(){
    return 0;
}

File: homework2/Saytam_Mishra/q5_stack.java
package homework2.Saytam_Mishra;

class Stack {
    class Node {
        int data;
        Node next;

        public Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    private Node head;

    public int top() { // O(1) time
        if(head == null) {
            throw new IllegalStateException("Stack is empty");
        }
        return head.data;
    }

    public void push(int x) { // O(1) time
        Node newNode = new Node(x);

        newNode.next = head;
        head = newNode;
    }

    public int pop() { // O(1) time
        if(isEmpty()) {
            throw new IllegalStateException("Stack is empty");
        } else {
            int delete = head.data;
            head = head.next;
            
            return delete;
        }
    }

    public boolean isEmpty() { // O(1) time
        return head == null;
    }
}

public class q5_stack {
    public static void main(String[] args) {
        Stack myStack = new Stack();

        myStack.push(10);
        myStack.push(20);
        System.out.println("Top element is: " + myStack.top());

        int val = myStack.pop();
        System.out.println("Popped: " + val);
        System.out.println("New top element is: " + myStack.top());
    }
}

// Time taken - 4m 20s